src/NServiceBus.Core/Recoverability/FirstLevelRetries/FailureInfoStorage.cs

```diff
@@ -0,0 +1,134 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Implements an in memory storage for message processing failures. Each failure contains a number of failed
+    /// processing attempts and the last captured exception.
+    /// </summary>
+    /// <remarks>
+    /// The data structure has fixed size configurable by a constructor parameter. When the data structure reaches
+    /// its maximum size, the least recently used (LRU) message processing failure is removed from the storage.
+    /// </remarks>
+    class FailureInfoStorage
```

Comment:
Also not convinced that we should reuse this one for the TM, I'd say we try to make the TM recoverability completely autonomous?